src/type-data-frame.c

```diff
@@ -59,29 +60,63 @@ SEXP vctrs_new_data_frame(SEXP args) {
     Rf_errorcall(R_NilValue, "`x` must be a list");
   }
 
+  bool has_rownames = false;
+
+  SEXP out = PROTECT(r_maybe_duplicate(x));
+
   R_len_t size = df_size_from_list(x, n);
 
-  if (attrib != R_NilValue) {
-    x = r_maybe_duplicate(x);
-    SET_ATTRIB(x, attrib);
+  for (SEXP attr_in = attrib; attr_in != R_NilValue; attr_in = CDR(attr_in)) {
+    SEXP tag = TAG(attr_in);
```

Comment:
Should still check for `tag == R_ClassSymbol` and throw an error if supplied.

(Or at least `continue` to skip the `setAttrib()`, because right now the class in dots is ignored if `cls` is `NULL`, but not otherwise. But I think it's probably better to make it an error as mentioned earlier.)

Reply:
How can a `"class"` sneak into the dots when `class` is part of the formals?

Reply:
oops sorry, I thought we had dynamic dots